### DISCUSSION TITLE
Window covering stuck on "Opening" state in Home app

### DISCUSSION THREAD

**Discussion by skarrelaar** [HomeSpan/HomeSpan#87] [General]
Hi! Let me start with a thank you for building this library. It's briliant. 

I am building some blinds that open and close with HomeKit, and so far everything is working. I have a stepper motor attached to my ESP32 that does everything I want it to. There is one problem though. When I adjust the position of the blinds in the Home app, the code runs fine, but when it is done the app gets stuck on "opening" as shown in the picture. This only happens when the shades are opened (at any percentage), when it's closed it shows up as if it's off and displays "closed".

![1CAA22FE-4473-4593-92F6-1C9BFDE2F659](https://user-images.githubusercontent.com/85135877/120326209-c6e48b80-c2e8-11eb-895b-a4935ccd7d11.jpeg)

 However, if I open the panel for the specific accessory it does show up with the correct position and it doesn't say "opening" as displayed in the picture below.

![B4E7E75D-A844-45CB-BA6D-3D3F670DCC1B](https://user-images.githubusercontent.com/85135877/120327543-28f1c080-c2ea-11eb-81a0-a5436446a37b.png)

I noticed that HAP specifies three required characteristics for window coverings: `Target position`, `current position`and `position state`. In the code I really only use `target position`, but as a test I made `position state` always `2`, which means "stopped". That did not work. I also tried making sure that `current positon` is the same value as `target position` at the end of the code (just before `update` returns `true`, but that didn't fix it either.

What could be the reason the Home app does this? Is there a fix for this? I could just continue without fixing this as the device does work, but it is a little annoying seeing that loading icon in my app all the time. 

Thanks!

**Comment by skarrelaar** [comment]
Ah I figured it out, apperantly I didn't go over the examples well enough. Found my answer in example 13.

**Comment by HomeSpan** [comment]
Glad to hear it's working correctly.  Can I ask ask what library and board you are using to drive your stepper motor?  I've been developing a driver for HomeSpan that uses the native Motor PWM functionality of the ESP32 itself.  You still need to use an H-Bridge, but all the stepper logic would be in the ESP32.  This is more for fun than practical use since there are already a variety of low-cost stepper-motor boards that contain both a driver chip and an H-bridge.

**Reply by skarrelaar** [reply]
I’m using a DRV8825 to drive it and I’m not using any library. I wanted to use AccelStepper at first but it took me a little too long to get it to work correctly haha, so now I just use the first method used in [This](https://www.makerguides.com/drv8825-stepper-motor-driver-arduino-tutorial/) blogpost.

**Comment by HomeSpan** [comment]
Thank for the info.  The method in the blogpost should cause the ESP32 to pause operations while the motor is moving, which means you would need to wait for the loop controlling the stepper motor to finish before HomeSpan resumes normal operation.  Is this what you observe?

**Reply by skarrelaar** [reply]
Yes I think that is happening, but so far I haven’t had any problems with it. Do you think I could run into anything using this method?

**Comment by HomeSpan** [comment]
This should not be a problem, though the Home App may complain about getting no response if there is a long delay while the motor is running and a response has not yet been sent back from HomeSpan.  I assume that if you use the Home App to raise the shade, you can't instruct the Home App to lower it until it has finished moving (and vice versa)?  In other words, you can't stop/reverse it while it is still moving?  If so, there are other ways of implementing the motor control that would run the motor in the background while returning control to HomeSpan to process any new commands.

**Reply by skarrelaar** [reply]
Yeah if I quickly change the values multiple times in a row it has to wait for the motor to fully work through all the commands, but that’s something I’m not planning on doing with my shades anyway haha. So far hadn’t had any trouble with HomeKit complaining about a lack of response. If I would run into something like that could it be solved by using ESP32’s multiple cores? Is that something you would recommend doing?

**Reply by blaisebarrette** [reply]
I also am curious about the use of the second core with HomeSpan.

**Reply by HomeSpan** [reply]
I am working through modifications that will make key aspects of HomeSpan "thread safe" so you can have a motor control thread operate in the background without causing HomeSpan to freeze.  See `homeSpan.autoPoll()` on the dev branch for a preview.
